cmd/tools/genrpcserverinterceptors/main.go

```diff
@@ -0,0 +1,287 @@
+
+	writeGrpcServerData(w, grpcServerT, `
+package logtags
+
+import (
+{{- range .Imports}}
+	{{printf "%q" .}}
+{{- end}}
+	"go.temporal.io/server/common/log/tag"
+)
+
+func (wt *WorkflowTags) extractFrom{{.Server}}Request(req any) []tag.Tag {
+	switch r := req.(type) {
+	{{- range .Requests}}
+	case {{.Type}}:
+	{{- if or .TaskTokenGetter .WorkflowIdGetter .RunIdGetter}}
+		{{- if .TaskTokenGetter}}
+		return wt.fromTaskToken(r.{{ .TaskTokenGetter}})
+		{{- else}}
+		return []tag.Tag{
+		{{if .WorkflowIdGetter}}	tag.WorkflowID(r.{{.WorkflowIdGetter}}),
+		{{end -}}
+		{{if .RunIdGetter}}	tag.WorkflowRunID(r.{{.RunIdGetter}}),
+		{{end -}}
+		}
+		{{- end}}
+	{{- else}}
+		return nil
+	{{- end -}}
+	{{- end}}
+	default:
+		return nil
+	}
+}
+`)
+}
+
+func writeGrpcServerData(w io.Writer, grpcServerT reflect.Type, tmpl string) {
+	sd := grpcServerData{
+		Server:  grpcServerT.Name(),
+		Imports: []string{grpcServerT.PkgPath()},
+	}
+
+	for i := 0; i < grpcServerT.NumMethod(); i++ {
+		rpcT := grpcServerT.Method(i).Type
+		if rpcT.NumIn() < 2 {
+			continue
+		}
+		requestT := rpcT.In(1) // Assume request is always the second parameter.
+
+		rd := workflowTagGetters(requestT, 0)
+		rd.Type = requestT.String()
+		// processOverrides(requestT, &rd)
+		sd.Requests = append(sd.Requests, rd)
+	}
+
+	fatalIfErr(template.Must(template.New("code").Parse(tmpl)).Execute(w, sd))
+}
+
+//nolint:revive // cognitive complexity 37 (> max enabled 25)
+func workflowTagGetters(requestT reflect.Type, depth int) requestData {
+	rd := requestData{}
+	if depth > maxRequestDepth {
+		return rd
+	}
+
+	switch {
+	case requestT.AssignableTo(executionGetterT):
+		rd.WorkflowIdGetter = "GetExecution().GetWorkflowId()"
+		rd.RunIdGetter = "GetExecution().GetRunId()"
+	case requestT.AssignableTo(workflowExecutionGetterT):
+		rd.WorkflowIdGetter = "GetWorkflowExecution().GetWorkflowId()"
+		rd.RunIdGetter = "GetWorkflowExecution().GetRunId()"
+	case requestT.AssignableTo(taskTokenGetterT):
+		for _, ert := range excludeTaskTokenTypes {
+			if requestT.AssignableTo(ert) {
+				return rd
+			}
+		}
+		rd.TaskTokenGetter = "GetTaskToken()"
+	default:
+		// Might be one of these, both, or neither.
+		if requestT.AssignableTo(workflowIdGetterT) {
+			rd.WorkflowIdGetter = "GetWorkflowId()"
+		}
+		if requestT.AssignableTo(runIdGetterT) {
+			rd.RunIdGetter = "GetRunId()"
+		}
+	}
+
+	for fieldNum := 0; fieldNum < requestT.Elem().NumField(); fieldNum++ {
+		// Iterates over fields in order they defined in proto file, not proto index.
+		// Order is important because the first match wins.
+		nestedRequest := requestT.Elem().Field(fieldNum)
+
+		if nestedRequest.Type.Kind() != reflect.Ptr {
+			continue
+		}
+		if nestedRequest.Type.Elem().Kind() != reflect.Struct {
+			continue
+		}
+		if !fieldNameRegex.MatchString(nestedRequest.Name) {
+			continue
+		}
+
+		nestedRd := workflowTagGetters(nestedRequest.Type, depth+1)
```

Comment:
also, I guess you know, you can cut off the recursion and iteration when all fields are nonzero. but I imagine the runtime is instant anyway

Reply:
I had it, then removed. I am brining it back.